extensions/amp-gwd-animation/0.1/amp-gwd-animation.js

```diff
@@ -0,0 +1,228 @@
+          gwdPageDeck, 'slideChange', setCurrentPageActionDef);
+    }
+
+    // Register handlers for supported actions.
+    for (const name in ACTION_IMPL_ARGS) {
+      this.registerAction(name, this.createAction_(name));
+    }
+  }
+
+  /**
+   * Returns a registrable AMP action function which invokes the corresponding
+   * GWD runtime method with arguments extracted from the invocation object
+   * (@see getActionImplArgs).
+   * @param {string} actionName Name of the action to invoke (currently
+   *     identical to the corresponding service method).
+   * @return {!function(!../../../src/service/action-impl.ActionInvocation)}
+   * @private
+   */
+  createAction_(actionName) {
+    return invocation => {
+      const service = user().assert(
+          getServiceForDoc(this.getAmpDoc(), GWD_SERVICE_NAME),
+          'Cannot execute action because the GWD service is not registered.');
+
+      const actionArgs = getActionImplArgs(actionName, invocation);
+      service[actionName].apply(service, actionArgs);
+    };
+  }
+
+  /**
+   * Handles `gwd.timelineEvent` events dispatched by the runtime service, which
+   * indicate that the animation has reached a GWD event marker. Triggers a
+   * custom event via the doc action service.
+   * @param {!Event} event The `gwd.timelineEvent` event.
+   * @private
+   */
+  onGwdTimelineEvent_(event) {
+    Services.actionServiceForDoc(this.getAmpDoc()).trigger(
+        this.element,
+        `${this.timelineEventPrefix_}${event.detail.eventName}`,
+        event.detail,
+        ActionTrust.HIGH);
+  }
+
+  /** @override */
+  detachedCallback() {
+    this.getAmpDoc().win.removeEventListener(
+        GWD_TIMELINE_EVENT, this.boundOnGwdTimelineEvent_, true);
+    return true;
+  }
+}
+
+/**
+ * Modifies the given element's `on` attribute to include the given event and
+ * action handler definition. (This is currently the only mechanism by which
+ * an AMP event handler can be programmatically added.)
+ * @param {!Element} element
+ * @param {string} event
+ * @param {string} actionStr e.g., `someDiv.hide`
+ * @private Visible for testing.
+ */
+export const insertEventActionBinding = function(element, event, actionStr) {
+  const currentOnAttrVal = element.getAttribute('on') || '';
+  const eventPrefix = `${event}:`;
+  const eventDefIndex = currentOnAttrVal.indexOf(eventPrefix);
+  let newOnAttrVal;
+
+  if (eventDefIndex != -1) {
+    // Some actions already defined for this event. Splice in the new action.
+    const actionsDefIndex = eventDefIndex + eventPrefix.length;
+
+    newOnAttrVal =
+        currentOnAttrVal.substr(0, actionsDefIndex) +
+        actionStr + ',' +
+        currentOnAttrVal.substr(actionsDefIndex);
+  } else {
+    // No actions defined yet for this event. Append the new action.
+    newOnAttrVal =
+        currentOnAttrVal +
+        (currentOnAttrVal ? ';' : '') +
+        eventPrefix +
+        actionStr;
+  }
+
+  element.setAttribute('on', newOnAttrVal);
```

Comment:
Just fyi, you might be also able to use `actionService.installActionHandler(...)`

Reply:
Does that entirely replace the handler for all actions on the element? I think I need the default AMP action handler to remain, since it may have other event/action definitions that I don't want to have to handle myself, but simply add an additional action def to the list. I couldn't find another way to do it except to manually manipulate the `on` attribute value. Let me know if there's a better way.

Reply:
This is a bit racy because actions could be cached. Can we maybe change this a bit via calling a method on `actionService` that will do `element.setAttribute('on', ...)` but will also reset the cached value?

Reply:
Added addAction method as discussed.